core/src/main/java/com/linecorp/armeria/common/stream/TimeoutSubscriber.java

```diff
@@ -0,0 +1,134 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.TimeoutException;
+
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.ScheduledFuture;
+
+final class TimeoutSubscriber<T> implements Runnable, Subscriber<T> {
+    private static final String TIMEOUT_MESSAGE = "Stream timed out after %d ms (timeout mode: %s)";
+    private final Subscriber<? super T> delegate;
+    private final EventExecutor executor;
+    private final StreamTimeoutMode timeoutMode;
+    private final Duration timeoutDuration;
+    private final long timeoutNanos;
+    private ScheduledFuture<?> timeoutFuture;
+    private Subscription subscription;
+    private long lastOnNextTimeNanos;
+    private boolean isTerminated;
+
+    TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration,
+                      StreamTimeoutMode timeoutMode) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.executor = requireNonNull(executor, "executor");
+        this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+        timeoutNanos = timeoutDuration.toNanos();
+        this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+    }
+
+    private ScheduledFuture<?> createTimeoutSchedule(long delay, TimeUnit unit) {
+        return executor.schedule(this, delay, unit);
+    }
+
+    private boolean attemptTerminate() {
+        if (isTerminated) {
+            return false;
+        }
+        isTerminated = true;
+        return true;
+    }
+
+    public void cancelSchedule() {
+        if (!timeoutFuture.isCancelled()) {
+            timeoutFuture.cancel(false);
+        }
+    }
+
+    @Override
+    public void run() {
+        if (timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+            final long currentTimeNanos = System.nanoTime();
+            final long elapsedNanos = currentTimeNanos - lastOnNextTimeNanos;
+
+            if (elapsedNanos < timeoutNanos) {
+                final long delayNanos = timeoutNanos - elapsedNanos;
+                timeoutFuture = createTimeoutSchedule(delayNanos, TimeUnit.NANOSECONDS);
+                return;
+            }
+        }
+        if (!attemptTerminate()) {
```

Comment:
Let's move this to the very top of this method.

Reply:
Oops please forget about this comment. 😓

Reply:
Let's do
```suggestion
        completed = true;
```
because `completed` is always `false` when this method is called.